File: go/email_conformance_test.go
package haisdk

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"testing"
)

// conformanceFixture mirrors the JSON structure of fixtures/email_conformance.json.
type conformanceFixture struct {
	ContentHashGolden struct {
		Vectors []struct {
			Name         string `json:"name"`
			Subject      string `json:"subject"`
			Body         string `json:"body"`
			Attachments  []struct {
				Filename    string `json:"filename"`
				ContentType string `json:"content_type"`
				DataUTF8    string `json:"data_utf8"`
			} `json:"attachments"`
			ExpectedHash string `json:"expected_hash"`
		} `json:"vectors"`
	} `json:"content_hash_golden"`

	VerificationResultV2Schema struct {
		RequiredFields   map[string]string `json:"required_fields"`
		FieldStatusValues []string          `json:"field_status_values"`
	} `json:"verification_result_v2_schema"`

	APIContracts struct {
		SignEmail struct {
			Method             string `json:"method"`
			Path               string `json:"path"`
			RequestContentType string `json:"request_content_type"`
		} `json:"sign_email"`
		VerifyEmail struct {
			Method             string `json:"method"`
			Path               string `json:"path"`
			RequestContentType string `json:"request_content_type"`
		} `json:"verify_email"`
		SendEmail struct {
			ExcludedFields []string `json:"excluded_fields"`
		} `json:"send_email"`
	} `json:"api_contracts"`

	MockVerifyResponse struct {
		JSON json.RawMessage `json:"json"`
	} `json:"mock_verify_response"`
}

func loadConformanceFixture(t *testing.T) conformanceFixture {
	t.Helper()
	data, err := os.ReadFile(filepath.Join("..", "fixtures", "email_conformance.json"))
	if err != nil {
		t.Fatalf("read email_conformance.json: %v", err)
	}
	var f conformanceFixture
	if err := json.Unmarshal(data, &f); err != nil {
		t.Fatalf("unmarshal conformance fixture: %v", err)
	}
	return f
}

// ---------------------------------------------------------------------------
// EmailVerificationResultV2 structural conformance
// ---------------------------------------------------------------------------

func TestConformanceMockVerifyResponseDeserialization(t *testing.T) {
	f := loadConformanceFixture(t)

	var result EmailVerificationResultV2
	if err := json.Unmarshal(f.MockVerifyResponse.JSON, &result); err != nil {
		t.Fatalf("failed to deserialize mock verify response into EmailVerificationResultV2: %v", err)
	}

	// Verify all required fields were populated.
	if result.JacsID != "conformance-test-agent-001" {
		t.Fatalf("jacs_id mismatch: %s", result.JacsID)
	}
	if !result.Valid {
		t.Fatal("expected valid=true")
	}
	if result.Algorithm != "ed25519" {
		t.Fatalf("algorithm mismatch: %s", result.Algorithm)
	}
	if result.ReputationTier != "established" {
		t.Fatalf("reputation_tier mismatch: %s", result.ReputationTier)
	}
	if result.DNSVerified == nil || !*result.DNSVerified {
		t.Fatal("expected dns_verified=true")
	}
	if result.Error != nil {
		t.Fatalf("expected error=nil, got %q", *result.Error)
	}

	// Verify field_results.
	if len(result.FieldResults) != 4 {
		t.Fatalf("expected 4 field_results, got %d", len(result.FieldResults))
	}
	if result.FieldResults[0].Field != "subject" || result.FieldResults[0].Status != FieldStatusPass {
		t.Fatalf("unexpected field_results[0]: %+v", result.FieldResults[0])
	}
	if result.FieldResults[3].Field != "date" || result.FieldResults[3].Status != FieldStatusModified {
		t.Fatalf("unexpected field_results[3]: %+v", result.FieldResults[3])
	}

	// Verify chain.
	if len(result.Chain) != 1 {
		t.Fatalf("expected 1 chain entry, got %d", len(result.Chain))
	}
	if result.Chain[0].Signer != "agent@hai.ai" || result.Chain[0].JacsID != "conformance-test-agent-001" {
		t.Fatalf("unexpected chain[0]: %+v", result.Chain[0])
	}
	if !result.Chain[0].Valid || result.Chain[0].Forwarded {
		t.Fatalf("chain[0] should be valid=true, forwarded=false")
	}
}

// ---------------------------------------------------------------------------
// FieldStatus enum conformance
// ---------------------------------------------------------------------------

func TestConformanceFieldStatusValues(t *testing.T) {
	f := loadConformanceFixture(t)
	expectedValues := f.VerificationResultV2Schema.FieldStatusValues

	goValues := map[FieldStatus]bool{
		FieldStatusPass:         true,
		FieldStatusModified:     true,
		FieldStatusFail:         true,
		FieldStatusUnverifiable: true,
	}

	for _, expected := range expectedValues {
		fs := FieldStatus(expected)
		if !goValues[fs] {
			t.Fatalf("FieldStatus %q from conformance fixture is not defined in Go SDK", expected)
		}
	}

	if len(goValues) != len(expectedValues) {
		t.Fatalf("Go SDK has %d FieldStatus values but conformance fixture has %d",
			len(goValues), len(expectedValues))
	}
}

// ---------------------------------------------------------------------------
// API contract conformance: SignEmail
// ---------------------------------------------------------------------------

func TestConformanceSignEmailAPIContract(t *testing.T) {
	f := loadConformanceFixture(t)

	var gotMethod, gotPath, gotContentType string
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotMethod = r.Method
		gotPath = r.URL.Path
		gotContentType = r.Header.Get("Content-Type")
		w.Header().Set("Content-Type", "message/rfc822")
		_, _ = w.Write([]byte("signed"))
	}))
	defer srv.Close()

	cl, _ := newTestClient(t, srv.URL)
	_, err := cl.SignEmail(context.Background(), []byte("raw email"))
	if err != nil {
		t.Fatalf("SignEmail: %v", err)
	}

	if gotMethod != f.APIContracts.SignEmail.Method {
		t.Fatalf("expected method %s, got %s", f.APIContracts.SignEmail.Method, gotMethod)
	}
	if gotPath != f.APIContracts.SignEmail.Path {
		t.Fatalf("expected path %s, got %s", f.APIContracts.SignEmail.Path, gotPath)
	}
	if gotContentType != f.APIContracts.SignEmail.RequestContentType {
		t.Fatalf("expected content-type %s, got %s", f.APIContracts.SignEmail.RequestContentType, gotContentType)
	}
}

// ---------------------------------------------------------------------------
// API contract conformance: VerifyEmail
// ---------------------------------------------------------------------------

func TestConformanceVerifyEmailAPIContract(t *testing.T) {
	f := loadConformanceFixture(t)

	var gotMethod, gotPath, gotContentType string
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotMethod = r.Method
		gotPath = r.URL.Path
		gotContentType = r.Header.Get("Content-Type")
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write(f.MockVerifyResponse.JSON)
	}))
	defer srv.Close()

	cl, _ := newTestClient(t, srv.URL)
	_, err := cl.VerifyEmail(context.Background(), []byte("raw email"))
	if err != nil {
		t.Fatalf("VerifyEmail: %v", err)
	}

	if gotMethod != f.APIContracts.VerifyEmail.Method {
		t.Fatalf("expected method %s, got %s", f.APIContracts.VerifyEmail.Method, gotMethod)
	}
	if gotPath != f.APIContracts.VerifyEmail.Path {
		t.Fatalf("expected path %s, got %s", f.APIContracts.VerifyEmail.Path, gotPath)
	}
	if gotContentType != f.APIContracts.VerifyEmail.RequestContentType {
		t.Fatalf("expected content-type %s, got %s", f.APIContracts.VerifyEmail.RequestContentType, gotContentType)
	}
}

// ---------------------------------------------------------------------------
// API contract conformance: SendEmail excluded fields
// ---------------------------------------------------------------------------

func TestConformanceSendEmailExcludesClientSigningFields(t *testing.T) {
	f := loadConformanceFixture(t)

	var gotBody map[string]interface{}
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		body, _ := io.ReadAll(r.Body)
		_ = json.Unmarshal(body, &gotBody)
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"message_id":"msg-conf","status":"sent"}`))
	}))
	defer srv.Close()

	cl, _ := newTestClient(t, srv.URL)
	_, err := cl.SendEmail(context.Background(), "bob@hai.ai", "Test", "Body")
	if err != nil {
		t.Fatalf("SendEmail: %v", err)
	}

	for _, excluded := range f.APIContracts.SendEmail.ExcludedFields {
		if _, ok := gotBody[excluded]; ok {
			t.Fatalf("SendEmail payload must not contain %q (server handles signing)", excluded)
		}
	}
}

// ---------------------------------------------------------------------------
// Error type conformance
// ---------------------------------------------------------------------------

func TestConformanceErrorTypeSentinels(t *testing.T) {
	// Verify all sentinel errors exist and have the expected string representation.
	sentinels := map[string]error{
		"EMAIL_NOT_ACTIVE":  ErrEmailNotActive,
		"RECIPIENT_NOT_FOUND": ErrRecipientNotFound,
		"RATE_LIMITED":       ErrEmailRateLimited,
	}

	for code, sentinel := range sentinels {
		if sentinel == nil {
			t.Fatalf("sentinel for %q is nil", code)
		}
		if sentinel.Error() == "" {
			t.Fatalf("sentinel for %q has empty error string", code)
		}
	}
}

File: node/tests/email-conformance.test.ts
import { afterEach, describe, expect, it, vi } from 'vitest';
import { readFileSync } from 'node:fs';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { HaiClient } from '../src/client.js';
import { FFIClientAdapter } from '../src/ffi-client.js';
import { generateTestKeypair as generateKeypair } from './setup.js';
import { createMockFFI } from './ffi-mock.js';
import {
  EmailNotActiveError,
  RecipientNotFoundError,
  RateLimitedError,
} from '../src/errors.js';
import type {
  EmailVerificationResultV2,
  FieldStatus,
} from '../src/types.js';

// ---------------------------------------------------------------------------
// Fixture loading
// ---------------------------------------------------------------------------

interface ConformanceFixture {
  verification_result_v2_schema: {
    required_fields: Record<string, string>;
    field_status_values: string[];
  };
  api_contracts: {
    sign_email: { method: string; path: string; request_content_type: string };
    verify_email: { method: string; path: string; request_content_type: string };
    send_email: { excluded_fields: string[] };
  };
  mock_verify_response: { json: Record<string, unknown> };
}

function loadConformanceFixture(): ConformanceFixture {
  const here = dirname(fileURLToPath(import.meta.url));
  const fixturePath = resolve(here, '../../fixtures/email_conformance.json');
  return JSON.parse(readFileSync(fixturePath, 'utf-8')) as ConformanceFixture;
}

async function makeClient(baseUrl: string): Promise<HaiClient> {
  const keypair = generateKeypair();
  const client = await HaiClient.fromCredentials('test-agent-001', keypair.privateKeyPem, { url: baseUrl, privateKeyPassphrase: 'keygen-password' });
  client.setAgentEmail('test@hai.ai');
  return client;
}

// ---------------------------------------------------------------------------
// EmailVerificationResultV2 structural conformance
// ---------------------------------------------------------------------------

describe('email conformance: mock verify response deserialization', () => {
  const fixture = loadConformanceFixture();

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it('deserializes mock response into EmailVerificationResultV2 via verifyEmail', async () => {
    const mockJson = fixture.mock_verify_response.json;

    const client = await makeClient('https://mock.hai.ai');
    const verifyEmailRawMock = vi.fn(async () => mockJson);
    client._setFFIAdapter(createMockFFI({ verifyEmailRaw: verifyEmailRawMock }));

    const result: EmailVerificationResultV2 = await client.verifyEmail('raw email content');

    expect(result.valid).toBe(true);
    expect(result.jacsId).toBe('conformance-test-agent-001');
    expect(result.algorithm).toBe('ed25519');
    expect(result.reputationTier).toBe('established');
    expect(result.dnsVerified).toBe(true);
    expect(result.error).toBeNull();

    // field_results
    expect(result.fieldResults).toHaveLength(4);
    expect(result.fieldResults[0].field).toBe('subject');
    expect(result.fieldResults[0].status).toBe('pass');
    expect(result.fieldResults[3].field).toBe('date');
    expect(result.fieldResults[3].status).toBe('modified');

    // chain
    expect(result.chain).toHaveLength(1);
    expect(result.chain[0].signer).toBe('agent@hai.ai');
    expect(result.chain[0].jacsId).toBe('conformance-test-agent-001');
    expect(result.chain[0].valid).toBe(true);
    expect(result.chain[0].forwarded).toBe(false);

    // agent_status and benchmarks_completed (TASK_012)
    expect(result.agentStatus).toBe('active');
    expect(result.benchmarksCompleted).toEqual(['free_chaotic']);
  });
});

// ---------------------------------------------------------------------------
// FieldStatus enum conformance
// ---------------------------------------------------------------------------

describe('email conformance: FieldStatus values', () => {
  const fixture = loadConformanceFixture();

  it('all fixture field_status_values are valid FieldStatus literals', () => {
    const validStatuses: FieldStatus[] = ['pass', 'modified', 'fail', 'unverifiable'];
    for (const val of fixture.verification_result_v2_schema.field_status_values) {
      expect(validStatuses).toContain(val);
    }
    expect(fixture.verification_result_v2_schema.field_status_values).toHaveLength(validStatuses.length);
  });
});

// ---------------------------------------------------------------------------
// Raw-email FFI contract: signing returns base64, verification returns JSON.
// ---------------------------------------------------------------------------

const rawEmailCases = [
  {
    name: 'UTF-8 string',
    rawEmail: 'From: alice@example.test\r\nSubject: café\r\n\r\nHello ☃\r\n',
  },
  {
    name: 'binary Buffer',
    rawEmail: Buffer.concat([
      Buffer.from('From: alice@example.test\r\nContent-Transfer-Encoding: binary\r\n\r\n'),
      Buffer.from([0x00, 0xc3, 0xa9, 0xff, 0x0d, 0x0a]),
    ]),
  },
];

describe('email conformance: signEmail raw-email FFI contract', () => {

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it.each(rawEmailCases)('signs $name through signEmailRaw without sending it', async ({ rawEmail }) => {
    const client = await makeClient('https://mock.hai.ai');
    const signedBytes = Buffer.concat([
      Buffer.from('Content-Type: multipart/mixed; boundary=signed\r\n\r\n'),
      Buffer.from([0x00, 0xff, 0x0d, 0x0a]),
    ]);
    const signEmailRawMock = vi.fn(async () => signedBytes.toString('base64'));
    const sendSignedEmailMock = vi.fn(async () => {
      throw new Error('signEmail must not send an email');
    });
    client._setFFIAdapter(createMockFFI({
      signEmailRaw: signEmailRawMock,
      sendSignedEmail: sendSignedEmailMock,
    }));

    const result = await client.signEmail(rawEmail);
    expect(signEmailRawMock).toHaveBeenCalledTimes(1);
    expect(signEmailRawMock).toHaveBeenCalledWith(Buffer.from(rawEmail).toString('base64'));
    expect(sendSignedEmailMock).not.toHaveBeenCalled();
    expect(result).toBeInstanceOf(Buffer);
    expect(result.equals(signedBytes)).toBe(true);
  });

  it('preserves the native signing result as a base64 string rather than parsing JSON', async () => {
    const inputB64 = Buffer.from('Subject: sign\r\n\r\nbody\r\n').toString('base64');
    const signedB64 = Buffer.from('Subject: signed\r\n\r\nbody\r\n').toString('base64');
    const signEmailRaw = vi.fn(async () => signedB64);
    // Exercise the real adapter conversion with a native boundary double.
    const adapter = Object.assign(Object.create(FFIClientAdapter.prototype), {
      native: { signEmailRaw },
    }) as FFIClientAdapter;

    await expect(adapter.signEmailRaw(inputB64)).resolves.toBe(signedB64);
    expect(signEmailRaw).toHaveBeenCalledTimes(1);
    expect(signEmailRaw).toHaveBeenCalledWith(inputB64);
  });

  it('maps native raw-email signing errors without returning empty bytes', async () => {
    const adapter = Object.assign(Object.create(FFIClientAdapter.prototype), {
      native: { signEmailRaw: vi.fn(async () => { throw new Error('NotFound: email not active'); }) },
    }) as FFIClientAdapter;

    await expect(adapter.signEmailRaw('')).rejects.toBeInstanceOf(EmailNotActiveError);
  });
});

// ---------------------------------------------------------------------------
// API contract conformance: VerifyEmail
// ---------------------------------------------------------------------------

describe('email conformance: verifyEmail raw-email FFI contract', () => {
  const fixture = loadConformanceFixture();

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it.each(rawEmailCases)('verifies exact $name bytes through verifyEmailRaw', async ({ rawEmail }) => {
    const client = await makeClient('https://mock.hai.ai');
    const verifyEmailRawMock = vi.fn(async () => fixture.mock_verify_response.json);
    const verifyDocumentMock = vi.fn(async () => {
      throw new Error('verifyEmail must not use JSON-document verification');
    });
    client._setFFIAdapter(createMockFFI({
      verifyEmailRaw: verifyEmailRawMock,
      verifyDocument: verifyDocumentMock,
    }));

    const result = await client.verifyEmail(rawEmail);
    expect(verifyEmailRawMock).toHaveBeenCalledTimes(1);
    expect(verifyEmailRawMock).toHaveBeenCalledWith(Buffer.from(rawEmail).toString('base64'));
    expect(verifyDocumentMock).not.toHaveBeenCalled();
    expect(result.valid).toBe(true);
  });

  it('decodes the native JSON verification response', async () => {
    const inputB64 = Buffer.from('Subject: signed\r\n\r\nbody\r\n').toString('base64');
    const verifyEmailRaw = vi.fn(async () => JSON.stringify(fixture.mock_verify_response.json));
    const adapter = Object.assign(Object.create(FFIClientAdapter.prototype), {
      native: { verifyEmailRaw },
    }) as FFIClientAdapter;

    await expect(adapter.verifyEmailRaw(inputB64)).resolves.toEqual(fixture.mock_verify_response.json);
    expect(verifyEmailRaw).toHaveBeenCalledTimes(1);
    expect(verifyEmailRaw).toHaveBeenCalledWith(inputB64);
  });
});

// ---------------------------------------------------------------------------
// API contract conformance: SendEmail excluded fields
// ---------------------------------------------------------------------------

describe('email conformance: sendEmail excluded fields', () => {
  const fixture = loadConformanceFixture();

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it('does not send client-side signing fields', async () => {
    let capturedOptions: Record<string, unknown> = {};

    const client = await makeClient('https://mock.hai.ai');
    const sendEmailMock = vi.fn(async (options: Record<string, unknown>) => {
      capturedOptions = options;
      return { message_id: 'msg-conf', status: 'sent' };
    });
    client._setFFIAdapter(createMockFFI({ sendEmail: sendEmailMock }));

    await client.sendEmail({ to: 'bob@hai.ai', subject: 'Test', body: 'Body' });

    for (const excluded of fixture.api_contracts.send_email.excluded_fields) {
      expect(capturedOptions).not.toHaveProperty(excluded);
    }
  });
});

// ---------------------------------------------------------------------------
// Error type conformance
// ---------------------------------------------------------------------------

describe('email conformance: error types', () => {
  it('all email error sentinel classes exist', () => {
    expect(EmailNotActiveError).toBeDefined();
    expect(RecipientNotFoundError).toBeDefined();
    expect(RateLimitedError).toBeDefined();

    // Verify they are constructable
    const e1 = new EmailNotActiveError('test');
    const e2 = new RecipientNotFoundError('test');
    const e3 = new RateLimitedError('test');
    expect(e1.message).toBe('test');
    expect(e2.message).toBe('test');
    expect(e3.message).toBe('test');
  });
});
